UI: Create BraveActionsContainer and add Shields extension button to it

petemill opened the pull request:
Fix https://github.com/brave/brave-browser/issues/668

Intention is for Shields 'extension' to feel built-in.
BraveActionsContainer hooks in to extension system in order to display and update the button icon when necessary.
Removes the Shields extension button from main extensions area (ToolbarActionsBar / BrowserActionsContainer).
Disable context menu for shields extension button in brave actions area since there are no valid actions.
Patches the content of a method at LocationBarView::Layout since it could be more harmful to copy that functionality to our subclass because it is complex and somewhat functional logic. Attempts to minimize the patching necessary by passing in a new argument to the method.

_This PR goes hand-in-hand with https://github.com/brave/brave-extension/pull/57 which controls the badge color, restricts the text from growing too wide (`"99+"`) and uses a vector icon for better rendering._ <-- **Now merged**

`BraveActionsContainer` will also contain the Brave Rewards button, which will be implemented in a subsequent PR, as either a button observing the Rewards Service, or another extension with a browser action.

A subsequent PR will implement a less-rounded Location Bar that will prevent cutting off the badge:

![image](https://user-images.githubusercontent.com/741836/44622423-e7947280-a86c-11e8-88ca-c7126b9d0109.png)

With this PR only it looks like this:
![image](https://user-images.githubusercontent.com/741836/44623011-4bbd3380-a879-11e8-8c21-9edbf130bbcc.png)


## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [ ] Added/updated tests for this change (for new code or code which already has tests).
- [ ] Ran `git rebase -i` to squash commits (if needed).
- [ ] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.

## Test Plan:


## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions

petemill commented:
This is now ready for review

simonhong commented:
Sorry for late. I'll review today.

simonhong commented on browser/ui/views/brave_actions/brave_actions_container.cc:
```diff
@@ -0,0 +1,195 @@
+    // in order to give space for the bubble.
+    shields_button_view_->SetPreferredSize(gfx::Size(32, 24));
+    // Add extension view after separator view
+    AddChildViewAt(shields_button_view_, 1);
```
`AddChildView()` adds newly added child to last.
```
void View::AddChildView(View* view) {
  if (view->parent_ == this)
    return;
  AddChildViewAt(view, child_count());
}
```

petemill replied:
Yes. I had made this explicit as index 1 since we will be adding another button which we will want at index 2. In other words the order of the buttons must always be the same. But yes, when we do add that button in a future PR, we won't be able to add it at index 2 if the shields hasn't loaded yet to index 1, so we'll need a better way of positioning.

simonhong commented on browser/ui/views/brave_actions/brave_actions_container.h:
```diff
@@ -0,0 +1,105 @@
+
+    bool should_hide_ = false;
+
+    ToolbarActionView* shields_button_view_ = nullptr;
```
How about using `std::unique_ptr` for |shields_button_view_|?  because this class owns it.
In chromium, using raw point means it doesn't own it.

petemill replied:
I copied this pattern from chromium, for example at https://cs.chromium.org/chromium/src/chrome/browser/ui/views/location_bar/location_bar_view.h?type=cs&q=LocationBarView&sq=package:chromium&g=0&l=456

Could you let me know what the difference is?

simonhong replied:
It is different. Child view you referenced is owned by parent view. That means child view is destroyed when parent view destroyed.
In our case, we delete it explicitly. That means deleting can happen from view tree or our `delete`.

petemill replied:
Re-reading what chromium does for dynamic extension browser action icons: https://cs.chromium.org/chromium/src/chrome/browser/ui/views/toolbar/browser_actions_container.cc?type=cs&sq=package:chromium&g=0&l=158-179

Will consider not `delete`ing the view manually and maybe just RemoveChildView, or using `shields_button_view_->set_owned_by_client()` as suggested below.

simonhong replied:
Yep, I think `RemoveChildView` is more safer than `delete`.

simonhong replied:
I read again and I think worried situation(double delete) would not be happened.
Explicit deleting is done when `OnExtensionUnloaded()` is called and implicit deleting is done when parent(this class) is destroyed.
When parent is destroyed first, `OnExtensionUnloaded()` cannot be called again.
When `OnExtnsionUnloaded()` is called first, button view is removed from parent and it is explicitly deleted.
So, I think current impl doesn't have problem. WDYT?

petemill replied:
Yes @simonhong after thinking again I think we should keep the `delete` here since after calling `RemoveChildView` the view will never get `delete`d by the old parent, so we will never free the memory. And we cannot re-use the `ToolbarActionView` for another instance of the extension.

simonhong commented on browser/ui/views/brave_actions/brave_actions_container.h:
```diff
@@ -0,0 +1,105 @@
+    extensions::ExtensionActionAPI* extension_action_api_;
+    extensions::ExtensionRegistry* extension_registry_;
+    extensions::ExtensionActionManager* extension_action_manager_;
+    ToolbarActionViewController* shields_view_controller_ = nullptr;
```
same as `shields_button_view_`.

simonhong replied:
I think we should consider using smart pointer to |shields_view_controller_|.
Or if we can guarantee `RemoveShields()` is always called before `BraveActionsContainer` is destroyed,
just adding DCHECK(!shields_view_controller_) in dtor would be sufficient.

simonhong commented on browser/ui/views/brave_actions/brave_actions_container.cc:
```diff
@@ -0,0 +1,195 @@
+        browser_, extension_action_manager_->GetExtensionAction(*brave_extension),
+                                                                        nullptr);
+    shields_button_view_ = new ToolbarActionView(
+                                                  shields_view_controller_, this);
```
Let's call `shields_button_view_->set_owned_by_client()` because parent doesn't own button view.
With this, parent doesn't delete it by accident.

petemill replied:
I'm confused about this - BraveActionsContainer is the parent, since we add it as a child, right? If the Location Bar is deleted, it's ok to delete this button, no?

simonhong replied:
Ah,,, I'm concerning there are two different delete path.(by tree view or by our `delete`).
That can cause some problem like double delete.

petemill replied:
oh i see, thanks. If I call that, does that mean I need to delete in destructor too?

petemill replied:
(FYI I was trying to avoid double delete by setting to `nullptr` but perhaps it can be deleted by tree view first?)

simonhong replied:
Right. I worried that situation.
So, smart pointer can handle it.

simonhong commented on browser/ui/views/brave_actions/brave_actions_container.cc:
```diff
@@ -0,0 +1,195 @@
+  if (shields_view_controller_)
+    shields_view_controller_->UpdateState();
+  // only show separator if we're showing any buttons
+  const bool visible = !should_hide_ && shields_button_view_;
```
How about checking children count() instead of checking |shields_button_view_| pointer?
In the near future, I assume that this view can have more buttons.
Because of this, I think we should not assume this view can have only one button.

petemill replied:
Interesting idea. My intention was to add the one other button (Brave Rewards) to this conditional, once the button is added. I think checking explicitly what we require to be visible is appropriate since we may have a changing number of visual-only elements, such as the separator.

simonhong commented on browser/ui/views/brave_actions/brave_actions_container.cc:
```diff
@@ -0,0 +1,195 @@
+  if (shields_button_view_) {
+    RemoveChildView(shields_button_view_);
+    delete shields_button_view_;
+    delete shields_view_controller_;
```
same as below comment (children count)

simonhong commented on browser/ui/views/location_bar/brave_location_bar_view.cc:
```diff
@@ -0,0 +1,66 @@
+
+#include "chrome/browser/ui/layout_constants.h"
+#include "chrome/browser/ui/views/location_bar/location_bar_view.h"
+#include "brave/browser/ui/views/brave_actions/brave_actions_container.h"
```
nit: order

simonhong commented on browser/ui/views/location_bar/brave_location_bar_view.cc:
```diff
@@ -0,0 +1,66 @@
+#include "chrome/browser/ui/location_bar/location_bar.h"
+#include "brave/browser/ui/location_bar/brave_location_bar.h"
+
+#define LocationBar BraveLocationBar
```
what is this for?

petemill replied:
We were previously overriding LocationBar, a subclass of LocationBarView via a chromium_src override in LocationBarView, but now we can do it in this class instead.

simonhong commented on browser/ui/views/brave_actions/brave_actions_container.h:
```diff
@@ -0,0 +1,105 @@
+#ifndef BRAVE_BROWSER_UI_VIEWS_BRAVE_ACTIONS_BRAVE_ACTIONS_CONTAINER_H_
+#define BRAVE_BROWSER_UI_VIEWS_BRAVE_ACTIONS_BRAVE_ACTIONS_CONTAINER_H_
+
+
```
nit: two empty line.

petemill commented on browser/ui/brave_actions/brave_action_icon_with_badge_image_source.cc:
```diff
@@ -0,0 +1,98 @@
+    int w = 0;
+    int h = 0;
+    gfx::FontList bigger_font =
+        base_font.Derive(1, 0, gfx::Font::Weight::NORMAL);
```
This was copied from chromium implementation. I agree that maybe it's better, though perhaps it's best to stay the same as chromium in order to compare against future versions.

petemill commented on browser/ui/brave_actions/brave_action_icon_with_badge_image_source.cc:
```diff
@@ -0,0 +1,98 @@
+
+  constexpr int kBadgeHeight = 12;
+  constexpr int kPadding = 2;
+  const int kTextHeight = 2 + kBadgeHeight - (kPadding * 2);
```
The numbers were generally rendering smaller than the height so I added some ability to bleed in to padding area. But it would be cleaner to just reduce padding on height to `1`, so thanks.

petemill commented on browser/ui/brave_actions/brave_action_icon_with_badge_image_source.cc:
```diff
@@ -0,0 +1,98 @@
+    gfx::FontList bigger_font =
+        base_font.Derive(1, 0, gfx::Font::Weight::NORMAL);
+    gfx::Canvas::SizeStringInt(utf16_text, bigger_font, &w, &h, 0,
+                               gfx::Canvas::NO_ELLIPSIS);
```
Didn't try - this was copied from chromium `icon_with_badge_image_source`

simonhong commented on patches/chrome-browser-ui-views-location_bar-location_bar_view.cc.patch:
```diff
@@ -0,0 +1,25 @@
+ 
++  if (right_most && right_most->visible())
++    trailing_decorations.AddDecoration(0,
++                              location_height + (vertical_padding * 2),
```
how about using `height()` instead of `location_height + (vertical_padding * 2)`?
I think both are same.

petemill commented on browser/ui/brave_actions/brave_action_icon_with_badge_image_source.cc:
```diff
@@ -0,0 +1,98 @@
+  // l, t, r, b
+  rect.Inset(kHorizontalPadding, kVerticalPadding, kHorizontalPadding, kVerticalPadding);
+  // Draw string with ellipsis if it does not fit
+  canvas->DrawStringRect(utf16_text, base_font, text_color, rect);
```
I don't think that will help with the vertical centering

petemill commented on browser/ui/brave_actions/shields_action_view_controller.cc:
```diff
@@ -0,0 +1,68 @@
+  // ExtensionActionRunner* action_runner =
+  //     ExtensionActionRunner::GetForWebContents(web_contents);
+  // bool has_blocked_actions = action_runner && action_runner->WantsToRun(extension());
+  // image_source->set_paint_blocked_actions_decoration(has_blocked_actions);
```
thanks

petemill commented on browser/ui/brave_actions/shields_action_view_controller.cc:
```diff
@@ -0,0 +1,68 @@
+#include "ui/gfx/paint_vector_icon.h"
+#include "ui/gfx/scoped_canvas.h"
+
+void ShieldsActionViewController::HideActivePopup() {
```
I tested a couple of variations of opening popups and popups from other extensions and in all cases it did the same thing as chrome or other extensions in the normal extension bar. So I thought this would be ok for now and save the complexity of passing the ToolbarActionBar through LocationBarView to BraveActionsContainer into ShieldsActionViewController. Any thoughts?

simonhong replied:
Ah, then Ok!

simonhong commented on browser/ui/views/location_bar/brave_location_bar_view.cc:
```diff
@@ -0,0 +1,66 @@
+gfx::Size BraveLocationBarView::CalculatePreferredSize() const {
+  gfx::Size min_size = LocationBarView::CalculatePreferredSize();
+  if (brave_actions_ && brave_actions_->visible()){
+    int brave_actions_min = brave_actions_->GetMinimumSize().width();
```
I think `GetLayoutConstant(LOCATION_BAR_ELEMENT_PADDING)` value should be also considered.

petemill commented on browser/ui/views/brave_actions/brave_actions_container.cc:
```diff
@@ -0,0 +1,195 @@
+void BraveActionsContainer::AddShields(
+                                const extensions::Extension* brave_extension) {
+  DCHECK(brave_extension);
+  VLOG(1) << "AddShields, was already loaded:" << (bool)shields_button_view_;
```
This log line is intentionally here so we can get a log when the extension is loaded or re-loaded, regardless of the value of the `if` conditional. I think the confusion here is the text `was already loaded:` is pointing to the `bool` value of that statement and not saying that it was already loaded.

simonhong replied:
I see!

petemill commented on browser/ui/views/brave_actions/brave_actions_container.cc:
```diff
@@ -0,0 +1,195 @@
+}
+
+void BraveActionsContainer::RemoveShields() {
+  VLOG(1) << "RemoveShields, was already loaded:" << (bool)shields_button_view_;
```
Same as `AddShields` method

petemill commented on browser/ui/views/brave_actions/brave_actions_container.cc:
```diff
@@ -0,0 +1,195 @@
+                                                  shields_view_controller_, this);
+    // Sets overall size of button but not image graphic. We set a large width
+    // in order to give space for the bubble.
+    shields_button_view_->SetPreferredSize(gfx::Size(32, 24));
```
noted

petemill commented:
ok I have addressed all the review comments - thanks @simonhong 

I've also introduced the following concept after talking with the design team:
- Previously, the badge text font-size grew to meet a target height
- Now we introduce a max-width to the badge, and first decrease the font-size until that width is met.
- If the text was always more narrow than the max-width, then we attempt to increase the font-size to match the target height, as we did before.

This ensures that the logo is not covered as much.

bbondy commented on chromium_src/chrome/browser/ui/toolbar/toolbar_actions_model_factory.cc:
```diff
@@ -0,0 +1,6 @@
+#include "chrome/browser/ui/toolbar/toolbar_actions_model.h"
+#include "brave/browser/ui/toolbar/brave_toolbar_actions_model.h"
+
+#define ToolbarActionsModel BraveToolbarActionsModel
```
Always #undef after using #define and doing the include.

bbondy replied:
This can't be used because the base class has a static function with that return type in the file you're directly patching.

bbondy commented on chromium_src/chrome/browser/ui/views/toolbar/toolbar_view.cc:
```diff
@@ -0,0 +1,6 @@
+#include "chrome/browser/ui/views/location_bar/location_bar_view.h"
+#include "brave/browser/ui/views/location_bar/brave_location_bar_view.h"
+
+#define LocationBarView BraveLocationBarView
```
ditto
